src/main.py

```diff
@@ -93,44 +97,101 @@ def main() -> None:
                 "Can't exclude a goal predicate!"
     else:
         preds = env.predicates
+    # Create the agent (approach).
     approach = create_approach(CFG.approach, preds, env.options, env.types,
                                env.action_space)
-    # If approach is learning-based, get training datasets and do learning,
-    # testing after each learning call. Otherwise, just do testing.
+    # If agent is learning-based, generate offline dataset, allow the agent
+    # to learn from it, and then proceed with the online learning loop. Test
+    # after each learning call. If agent is not learning-based, just test once.
     if approach.is_learning_based:
-        if CFG.load_approach:
+        train_tasks = env.get_train_tasks()
+        dataset, total_num_transitions = _generate_or_load_offline_dataset(
+            env, train_tasks)
+        learning_start = time.time()
+        if CFG.load_approach:  # we only save/load for initial offline learning
             approach.load()
-            results = _run_testing(env, approach)
-            _save_test_results(results, learning_time=0.0)
         else:
-            # Create dataset from training tasks, and let agent learn.
-            train_tasks = env.get_train_tasks()
-            dataset_filename = (
-                f"{CFG.env}__{CFG.offline_data_method}__{CFG.seed}.data")
-            dataset_filepath = os.path.join(CFG.data_dir, dataset_filename)
-            if CFG.load_data:
-                assert os.path.exists(dataset_filepath)
-                with open(dataset_filepath, "rb") as f:
-                    dataset = pkl.load(f)
-                print("\n\nLOADED DATASET")
-            else:
-                dataset = create_dataset(env, train_tasks)
-                print("\n\nCREATED DATASET")
-                os.makedirs(CFG.data_dir, exist_ok=True)
-                with open(dataset_filepath, "wb") as f:
-                    pkl.dump(dataset, f)
-            learning_start = time.time()
             approach.learn_from_offline_dataset(dataset)
-            learning_time = time.time() - learning_start
+        teacher = Teacher()
+        for _ in range(CFG.num_online_learning_cycles):
             results = _run_testing(env, approach)
-            _save_test_results(results, learning_time=learning_time)
+            _save_test_results(results, num_transitions=total_num_transitions,
+                               learning_time=(time.time()-learning_start))
```

Comment:
num transitions and learning time are probably just two of several metrics that we'll want to save. should we get ahead of that and make a Metrics thing here?